server/src/main/java/io/druid/segment/realtime/appenderator/Appenderator.java

```diff
@@ -61,8 +61,6 @@
    * This method may trigger a {@link #persistAll(Committer)} using the supplied Committer. If it does this, the
    * Committer is guaranteed to be *created* synchronously with the call to add, but will actually be used
    * asynchronously.
-   * <p>
-   * The add, clear, persist, persistAll, and push methods should all be called from the same thread.
```

Comment:
Thanks. I fixed the sync problem in committed metadata. 

Also, I changed AppenderatorDriver.publish() to not retry. Instead, Appenderator does retry itself only for pushing segments. This is because exceptions from other parts are usually IOException while persisting segments or NPE due to logic error, etc and I don't think these exceptions are recoverable by executing multiple times.